File: contracts/orbital_math/torus.py
"""Torus invariant helpers for the Orbital simulator."""

from __future__ import annotations

from .constants import PRECISION, ROOT_PRECISION, TOLERANCE
from .fixed_point import div_round, isqrt, mul_div_round, normalized_square, sqrt_n_root_scaled
from .polar import alpha_total, w_norm_sq_from_sums


def _boundary_sum_target(k_bound: int, n: int, sqrt_n_value: int | None = None) -> int:
    sqrt_n_value = sqrt_n_value or sqrt_n_root_scaled(n)
    return mul_div_round(k_bound, sqrt_n_value, ROOT_PRECISION)


def _boundary_residual(
    sum_x: int,
    sum_x_sq: int,
    n: int,
    s_bound: int,
    k_bound: int,
    sqrt_n_value: int | None = None,
) -> int:
    sum_target = _boundary_sum_target(k_bound, n, sqrt_n_value)
    w_total_sq = w_norm_sq_from_sums(sum_x, sum_x_sq, n)
    norm_error = div_round(w_total_sq - (s_bound * s_bound), PRECISION)
    plane_error = sum_x - sum_target
    if abs(plane_error) <= 1:
        return norm_error
    return norm_error + plane_error


def torus_residual(
    sum_x: int,
    sum_x_sq: int,
    n: int,
    r_int: int,
    s_bound: int,
    k_bound: int,
    sqrt_n_value: int | None = None,
) -> int:
    sqrt_n_value = sqrt_n_value or sqrt_n_root_scaled(n)
    if r_int == 0:
        return _boundary_residual(sum_x, sum_x_sq, n, s_bound, k_bound, sqrt_n_value)

    alpha_total_value = alpha_total(sum_x, n)
    alpha_int = alpha_total_value - k_bound
    w_total_norm = isqrt(w_norm_sq_from_sums(sum_x, sum_x_sq, n))
    w_int_norm = w_total_norm - s_bound
    r_int_sqrt_n = mul_div_round(r_int, sqrt_n_value, ROOT_PRECISION)
    diff_alpha = alpha_int - r_int_sqrt_n

    lhs = normalized_square(r_int)
    rhs = normalized_square(diff_alpha) + normalized_square(w_int_norm)
    return lhs - rhs


def verify_torus(
    sum_x: int,
    sum_x_sq: int,
    n: int,
    r_int: int,
    s_bound: int,
    k_bound: int,
    sqrt_n_value: int | None = None,
    tolerance: int = TOLERANCE,
) -> bool:
    sqrt_n_value = sqrt_n_value or sqrt_n_root_scaled(n)
    if r_int == 0:
        sum_target = _boundary_sum_target(k_bound, n, sqrt_n_value)
        if abs(sum_x - sum_target) > 1:
            return False
    residual = torus_residual(sum_x, sum_x_sq, n, r_int, s_bound, k_bound, sqrt_n_value)
    return abs(residual) <= tolerance

File: frontend1/lib/orbital-sdk/math/torus.ts
import { PRECISION, TOLERANCE } from "../constants";
import { abs, sqrt } from "./bigint-math";

/**
 * Compute the torus invariant residual.
 *
 * Returns a value in PRECISION units — matching the Python reference's
 * normalized_square convention and the contract's square_scaled / _TOLERANCE.
 * |residual| <= TOLERANCE means the state satisfies the invariant.
 *
 * @param sumX    Σxᵢ  (PRECISION-scaled math reserves)
 * @param sumXSq  Σxᵢ² (PRECISION²-scaled raw squares — NOT divided by PRECISION)
 */
export function torusInvariant(
  sumX: bigint,
  sumXSq: bigint,
  n: number,
  rInt: bigint,
  sBound: bigint,
  kBound: bigint,
  sqrtN: bigint,
  invSqrtN: bigint,
): bigint {
  const N = BigInt(n);

  // Boundary-only case (all ticks at their boundary, rInt = 0).
  // Invariant degenerates to two independent constraints:
  //   1. sumX ≈ kBound * sqrt(n)   (plane constraint)
  //   2. ‖w‖ ≈ sBound              (boundary sphere constraint)
  if (rInt === 0n) {
    const sumTarget = (kBound * sqrtN) / PRECISION;
    const wTotalSq = sumXSq - (sumX * sumX) / N;
    const wSq = wTotalSq >= 0n ? wTotalSq : 0n;
    // normError = (‖w‖² - sBound²) / PRECISION  (PRECISION units)
    const normError = (wSq - sBound * sBound) / PRECISION;
    const planeError = sumX - sumTarget;
    // If the plane is satisfied within 1 unit, only report the norm error.
    return abs(planeError) <= 1n ? normError : normError + planeError;
  }

  // ── Alpha decomposition ──────────────────────────────────────────────────
  // alphaTotal = Σxᵢ / sqrt(n)  (PRECISION units)
  const alphaTotal = (sumX * invSqrtN) / PRECISION;
  // alphaInt = alphaTotal minus the boundary contribution
  const alphaInt = alphaTotal - kBound;
  // rInt * sqrt(n)  (PRECISION units)
  const rIntSqrtN = (rInt * sqrtN) / PRECISION;
  // diff along the interior sphere's polar axis
  const diffAlpha = alphaInt - rIntSqrtN;

  // ── w-norm decomposition ─────────────────────────────────────────────────
  // ‖w‖² = Σxᵢ² - (Σxᵢ)²/n   (PRECISION² units)
  const wTotalSq = sumXSq - (sumX * sumX) / N;
  const wTotalNorm = wTotalSq >= 0n ? sqrt(wTotalSq) : 0n; // PRECISION units
  // Strip the boundary sphere contribution
  const wIntNorm = wTotalNorm - sBound;

  // ── Torus check: rInt² = diffAlpha² + wIntNorm²  (all normalized to PRECISION units) ──
  // normalizedSquare(v) = v² / PRECISION  (PRECISION units) — matches Python + contract
  const lhs = (rInt * rInt) / PRECISION;
  const rhs = (diffAlpha * diffAlpha) / PRECISION + (wIntNorm * wIntNorm) / PRECISION;

  return lhs - rhs;
}

/**
 * Returns true if the pool state satisfies the torus invariant within TOLERANCE.
 */
export function isValidState(
  sumX: bigint,
  sumXSq: bigint,
  n: number,
  rInt: bigint,
  sBound: bigint,
  kBound: bigint,
  sqrtN: bigint,
  invSqrtN: bigint,
  tolerance: bigint = TOLERANCE,
): boolean {
  return abs(torusInvariant(sumX, sumXSq, n, rInt, sBound, kBound, sqrtN, invSqrtN)) <= tolerance;
}
